rattestation/attestation.c

```diff
@@ -51,26 +60,184 @@ convert_bin_to_hex_new(const uint8_t *bin, int length)
+	memcpy(sig, resp->signature.data, resp->signature.len);
+
+	// Convert to hex string representation to directly compare with specified values
+	// in the configuration file
+	for (size_t i = 0; i < resp->n_pcr_values; i++) {
+		pcr_strings[i] = convert_bin_to_hex_new(resp->pcr_values[i]->value.data,
+							resp->pcr_values[i]->value.len);
+	}
+
+	DEBUG("Verifying Response...");
+
+	DEBUG("Hash Algorithm: SHA%d", config->halg * 8);
+
+	// The pcrs are configured as hex strings
+	size_t pcr_string_len = config->halg * 2;
+
+	for (size_t i = 0; i < config->n_pcr_values; i++) {
+		if (strlen(pcr_strings[i]) != pcr_string_len) {
+			ERROR("Length of configured PCR value %ld invalid (%ld, must be %ld)", i,
+			      strlen(pcr_strings[i]), pcr_string_len);
+			ret = false;
+			goto err;
+		}
+		if (strlen(config->pcr_values[i]->value) != pcr_string_len) {
+			ERROR("Length of received PCR value %ld invalid (%ld, must be %ld)", i,
+			      strlen(config->pcr_values[i]->value), pcr_string_len);
+			ret = false;
+			goto err;
+		}
+		if (strncmp(pcr_strings[i], config->pcr_values[i]->value, pcr_string_len)) {
+			DEBUG("PCR_%ld: %s - VERIFICATION FAILED", i, pcr_strings[i]);
+			ret = false;
+			goto err;
+		}
+		DEBUG("PCR_%ld: %s - VERIFICATION SUCCESSFUL", i, pcr_strings[i]);
+	}
+
+	// The quote is sent as a TPMS_ATTEST and has to be unmarshalled
+	TPMS_ATTEST tpms_attest;
+	TPM_RC rc = TSS_TPMS_ATTEST_Unmarshalu(&tpms_attest, &q, &quote_len);
+	if (rc != TPM_RC_SUCCESS) {
+		ERROR("TSS_TPMS_ATTEST_Unmarshalu returned error code %u\n", rc);
+		ret = false;
+		goto err;
+	}
+
+	// Nonce verification
+	int ret_nonce = memcmp(tpms_attest.extraData.t.buffer, nonce, nonce_len);
+
 	char *nonce_str = convert_bin_to_hex_new(nonce, nonce_len);
-	INFO("Response for Nonce %s, Response size=%zu", nonce_str,
-	     protobuf_c_message_get_packed_size((ProtobufCMessage *)resp));
+	char *rcv_nonce_str = convert_bin_to_hex_new(tpms_attest.extraData.t.buffer,
+						     tpms_attest.extraData.t.size);
+	DEBUG("Nonce (sent %s, received %s) - %s", nonce_str, rcv_nonce_str,
+	      ret_nonce ? "VERIFICATION FAILED" : "VERIFICATION SUCCESSFUL");
 	mem_free(nonce_str);
-	//TODO implement verification process
-	return true;
+	mem_free(rcv_nonce_str);
+
+	// The signature is sent as a TPMT_SIGNATURE and has to be unmarshalled
+	TPMT_SIGNATURE tpmt_signature;
+	rc = TSS_TPMT_SIGNATURE_Unmarshalu(&tpmt_signature, &s, &sig_len, false);
+	if (rc != TPM_RC_SUCCESS) {
+		ERROR("TSS_TPMT_SIGNATURE_Unmarshalu returned error code %u\n", rc);
+		// TODO Error Handling
+		ret = false;
+		goto err;
+	}
+
+	// Verify signature
+	ssl_init(false, NULL);
+
+	int retssl = ssl_verify_signature_from_buf((const char *)resp->certificate.data,
+						   tpmt_signature.signature.rsapss.sig.t.buffer,
+						   tpmt_signature.signature.rsapss.sig.t.size,
+						   resp->quoted.data, resp->quoted.len);
+	if (retssl != 0) {
+		// TODO error handling
```

Comment:
what do you mean with this todo? you do cleanup in jump label.
what is here left todo?

Reply:
Done (removed TODO)